internal/unionstore/pipelined_memdb.go

```diff
@@ -0,0 +1,332 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package unionstore
+
+import (
+	"context"
+	"math"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/errors"
+	tikverr "github.com/tikv/client-go/v2/error"
+	"github.com/tikv/client-go/v2/kv"
+	"github.com/tikv/client-go/v2/metrics"
+	"github.com/tikv/client-go/v2/util"
+)
+
+// PipelinedMemDB is a Store which contains
+//   - a mutable buffer for read and write
+//   - an immutable onflushing buffer for read
+//   - like MemDB, PipelinedMemDB also CANNOT be used concurrently
+type PipelinedMemDB struct {
+	// Like MemDB, this RWMutex only used to ensure memdbSnapGetter.Get will not race with
+	// concurrent memdb.Set, memdb.SetWithFlags, memdb.Delete and memdb.UpdateFlags.
+	sync.RWMutex
+	onFlushing              atomic.Bool
+	errCh                   chan error
+	flushFunc               FlushFunc
+	bufferBatchGetter       BufferBatchGetter
+	memDB                   *MemDB
+	flushingMemDB           *MemDB // the flushingMemDB is not wrapped by a mutex, because there is no data race in it.
+	len, size               int    // len and size records the total flushed and onflushing memdb.
+	generation              uint64
+	entryLimit, bufferLimit uint64
+}
+
+const (
+	// MinFlushKeys is the minimum number of keys to trigger flush.
+	// small batch can lead to poor performance and resource waste in random write workload.
+	// 10K batch size is large enough to get good performance with random write workloads in tests.
+	MinFlushKeys = 10000
+	// MinFlushSize is the minimum size of MemDB to trigger flush.
+	MinFlushSize = 16 * 1024 * 1024 // 16MB
+	// ForceFlushSizeThreshold is the threshold to force flush MemDB, which controls the max memory consumption of PipelinedMemDB.
+	ForceFlushSizeThreshold = 128 * 1024 * 1024 // 128MB
+)
+
+type pipelinedMemDBSkipRemoteBuffer struct{}
+
+// TODO: skip remote buffer by context is too obscure, add a new method to read local buffer.
+var pipelinedMemDBSkipRemoteBufferKey = pipelinedMemDBSkipRemoteBuffer{}
+
+// WithPipelinedMemDBSkipRemoteBuffer is used to skip reading remote buffer for saving RPC.
+func WithPipelinedMemDBSkipRemoteBuffer(ctx context.Context) context.Context {
+	return context.WithValue(ctx, pipelinedMemDBSkipRemoteBufferKey, struct{}{})
+}
+
+type FlushFunc func(uint64, *MemDB) error
+type BufferBatchGetter func(ctx context.Context, keys [][]byte) (map[string][]byte, error)
+
+func NewPipelinedMemDB(bufferBatchGetter BufferBatchGetter, flushFunc FlushFunc) *PipelinedMemDB {
+	return &PipelinedMemDB{
+		memDB:             newMemDB(),
+		errCh:             make(chan error, 1),
+		flushFunc:         flushFunc,
+		bufferBatchGetter: bufferBatchGetter,
+		generation:        0,
+		entryLimit:        math.MaxUint64,
+		bufferLimit:       math.MaxUint64,
```

Comment:
How about to set it to `memDB.bufferSizeLimit` to make sure they are consistent, although both of them are inited with `math.MaxUint64` by now.

Reply:
Done.